### PR TITLE
Fix DI's TimeToFirstServiceBenchmark for explicit engines

### PR DESCRIPTION
**Pull request by steveharter**
This fixes the DI benchmark for comparing against specific engines such as Expressions. This broke with https://github.com/dotnet/runtime/pull/38729. It is somewhat fragile; depends on reflection.

To run with the different engines, uncomment the `#define INTERNAL_DI`.

This also resets the general benchmarks in the normal case (without the `#define`) since there used to be 4 engines in the benchmarks, but they were all using the "default" engine. Now there's just the "default" engine shown for correctness.

Local 8.0 results with all engines:
|        Method |        Mode |         Mean |      Error |      StdDev |       Median |          Min |          Max |   Gen0 |   Gen1 | Allocated |
|-------------- |------------ |-------------:|-----------:|------------:|-------------:|-------------:|-------------:|-------:|-------:|----------:|
| BuildProvider |     Default |     1.101 us |  0.0287 us |   0.0331 us |     1.097 us |     1.049 us |     1.152 us | 0.5749 | 0.0127 |    5.9 KB |
|     Transient |     Default |     2.522 us |  0.0711 us |   0.0665 us |     2.532 us |     2.404 us |     2.613 us | 0.6902 |      - |   7.43 KB |
|        Scoped |     Default |     2.461 us |  0.0522 us |   0.0601 us |     2.469 us |     2.355 us |     2.554 us | 0.7850 | 0.0199 |   8.03 KB |
|     Singleton |     Default |     2.299 us |  0.0469 us |   0.0540 us |     2.325 us |     2.182 us |     2.363 us | 0.7202 | 0.0187 |   7.41 KB |
| BuildProvider |     Dynamic |     2.512 us |  0.0582 us |   0.0670 us |     2.517 us |     2.411 us |     2.635 us | 0.7544 | 0.0196 |   7.77 KB |
|     Transient |     Dynamic |     4.271 us |  0.0801 us |   0.0749 us |     4.305 us |     4.138 us |     4.365 us | 0.8973 | 0.0173 |   9.31 KB |
|        Scoped |     Dynamic |     4.242 us |  0.0830 us |   0.0776 us |     4.266 us |     4.117 us |     4.338 us | 0.9632 | 0.0166 |   9.91 KB |
|     Singleton |     Dynamic |     3.928 us |  0.0742 us |   0.0855 us |     3.939 us |     3.779 us |     4.080 us | 0.8985 | 0.0152 |   9.28 KB |
| BuildProvider | Expressions |     2.606 us |  0.0612 us |   0.0601 us |     2.611 us |     2.526 us |     2.690 us | 0.7392 |      - |   7.77 KB |
|     Transient | Expressions |    64.455 us |  7.2404 us |   7.4353 us |    61.173 us |    59.622 us |    87.811 us | 1.1574 | 0.7716 |  14.04 KB |
|        Scoped | Expressions | 1,610.246 us | 88.9287 us | 102.4104 us | 1,584.281 us | 1,495.730 us | 1,754.255 us |      - |      - |  58.61 KB |
|     Singleton | Expressions |     4.914 us |  1.1190 us |   1.1973 us |     4.449 us |     3.982 us |     8.656 us | 0.8929 |      - |   9.28 KB |
| BuildProvider |      ILEmit |     2.518 us |  0.0682 us |   0.0785 us |     2.535 us |     2.409 us |     2.670 us | 0.7537 | 0.0193 |   7.77 KB |
|     Transient |      ILEmit |    54.956 us |  0.7391 us |   0.6914 us |    55.058 us |    53.496 us |    56.056 us | 0.8883 | 0.6662 |  10.91 KB |
|        Scoped |      ILEmit | 1,097.178 us | 19.3399 us |  18.0906 us | 1,094.389 us | 1,065.541 us | 1,119.894 us |      - |      - |  23.84 KB |
|     Singleton |      ILEmit |     4.573 us |  0.6444 us |   0.6617 us |     4.291 us |     4.131 us |     6.693 us | 0.8794 |      - |   9.27 KB |
| BuildProvider |     Runtime |     2.116 us |  0.0420 us |   0.0483 us |     2.114 us |     2.032 us |     2.177 us | 0.6180 | 0.0086 |   6.38 KB |
|     Transient |     Runtime |     3.708 us |  0.0823 us |   0.0808 us |     3.718 us |     3.570 us |     3.843 us | 0.7380 |      - |   7.89 KB |
|        Scoped |     Runtime |     5.170 us |  1.5652 us |   1.8025 us |     4.084 us |     3.911 us |     8.943 us | 0.7876 |      - |   8.47 KB |
|     Singleton |     Runtime |     3.796 us |  0.0754 us |   0.0740 us |     3.786 us |     3.663 us |     3.912 us | 0.7427 |      - |   7.89 KB |

Here's what the benchmarks will look like by default since the `INTERNAL_DI` #define is commented out to avoid issues on other platforms.

|        Method |    Mode |     Mean |     Error |    StdDev |   Median |      Min |      Max |   Gen0 |   Gen1 | Allocated |
|-------------- |-------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|-------:|----------:|
| BuildProvider | Default | 1.116 us | 0.0183 us | 0.0179 us | 1.115 us | 1.089 us | 1.159 us | 0.5775 | 0.0088 |    5.9 KB |
|     Transient | Default | 2.248 us | 0.0229 us | 0.0214 us | 2.246 us | 2.219 us | 2.285 us | 0.7205 | 0.0180 |   7.43 KB |
|        Scoped | Default | 2.507 us | 0.0209 us | 0.0174 us | 2.502 us | 2.482 us | 2.545 us | 0.7784 | 0.0200 |   8.03 KB |
|     Singleton | Default | 2.279 us | 0.0434 us | 0.0385 us | 2.268 us | 2.239 us | 2.382 us | 0.7225 | 0.0181 |   7.41 KB |

### CODE REVIEW

File: src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs [hunk 1]
```diff
@@ -119,5 +104,90 @@ private class B
...
+                return services.BuildServiceProvider();
+            }
+
+            Assembly asm = typeof(ServiceProvider).Assembly;
+
+            Type serviceProviderEngineType = asm.GetType("Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine", throwOnError: true, ignoreCase: false);
+            if (serviceProviderEngineType == null)
+            {
+                throw new Exception($"Unable to find ServiceProviderEngine type.");
+            }
```
Comment (eerhardt) [inline src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs:121]: This Type doesn't appear to be used. Is this code necessary?

File: src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs [hunk 2]
```diff
@@ -119,5 +104,90 @@ private class B
...
+            return _mode switch
+            {
+                ServiceProviderMode.Dynamic => CreateInstance("Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine"),
+                ServiceProviderMode.Runtime => CreateInstance("Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeServiceProviderEngine"),
+                ServiceProviderMode.Expressions => CreateInstance("Microsoft.Extensions.DependencyInjection.ServiceLookup.ExpressionsServiceProviderEngine"),
+                ServiceProviderMode.ILEmit => CreateInstance("Microsoft.Extensions.DependencyInjection.ServiceLookup.ILEmitServiceProviderEngine"),
+                _ => throw new NotSupportedException()
+            };
+
+            ServiceProvider CreateInstance(string typeName)
```
Comment (eerhardt) [inline src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs:134]:
```suggestion
            ServiceProvider CreateInstance(string engineTypeName)
```

**Review by eerhardt — APPROVED** [review]

File: src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs (hunk 1 above)
Comment (steveharter) [reply src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs:121]: Removed. Thanks

File: src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs [hunk 3]
```diff
@@ -2,6 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+// Define INTERNAL_DI on platforms that support emit.
+#define INTERNAL_DI
```
Comment (buyaa-n) [inline src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/TimeToFirstServiceBenchmark.cs:6]: > To run with the different engines, uncomment the define.

NIT: Maybe you were intending to comment it by default?
```suggestion
//#define INTERNAL_DI
```

File: src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/ActivatorUtilitiesBenchmark.cs [hunk 4]
```diff
@@ -8,6 +8,77 @@
 
 namespace Microsoft.Extensions.DependencyInjection
 {
+    [BenchmarkCategory(Categories.Libraries)]
+    public class AA
```
Comment (buyaa-n) [inline src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/ActivatorUtilitiesBenchmark.cs:12]: NIT: A meaningful name could be useful

**Review by buyaa-n — APPROVED** [review]
Left an NIT, overall looks good

File: src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/ActivatorUtilitiesBenchmark.cs (hunk 4 above)
Comment (steveharter) [reply src/benchmarks/micro/libraries/Microsoft.Extensions.DependencyInjection/ActivatorUtilitiesBenchmark.cs]: Removed this; it was a temp benchmark to help diagnose an issue.

**Comment by steveharter** [comment]
Ci issues with `windows 22HE` look consistent with other recent PRs.

**Review by buyaa-n — APPROVED** [review]

**Pull request merged**